This file is part of mingw-cross-env.
See doc/index.html for further information.

Commits backported (cherry-picked) from upstream.
http://github.com/bagder/curl
Also contains mingw-cross-env specific fixes.

From 39817fb04ff2ea4a693a1e853ca5fbd469c2b531 Mon Sep 17 00:00:00 2001
From: Tim Harder <radhermit@gmail.com>
Date: Wed, 19 Oct 2011 10:08:56 +0200
Subject: [PATCH 1/3] gtls: only call gnutls_transport_set_lowat with
 <gnutls-2.12.0

The default lowat level for gnutls-2.12* is set to zero to avoid
unnecessary system calls and the gnutls_transport_set_lowat function has
been totally removed in >=gnutls-3 which causes build failures.

Therefore, the function shouldn't be used except for versions that
require it, <gnutls-2.12.0.
(cherry picked from commit 8036da870c5b413a83097b3486c58d13910a471a)
---
 lib/gtls.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/lib/gtls.c b/lib/gtls.c
index f75a815..7ca46c8 100644
--- a/lib/gtls.c
+++ b/lib/gtls.c
@@ -476,8 +476,10 @@ gtls_connect_step1(struct connectdata *conn,
   gnutls_transport_set_push_function(session, Curl_gtls_push);
   gnutls_transport_set_pull_function(session, Curl_gtls_pull);
 
+#if GNUTLS_VERSION_NUMBER < 0x020c00
   /* lowat must be set to zero when using custom push and pull functions. */
   gnutls_transport_set_lowat(session, 0);
+#endif
 
   /* This might be a reconnect, so we check for a session ID in the cache
      to speed up things */
-- 
1.7.7


From 878969d2e89751ff445b1f6cb6ac05ad0bc07eb6 Mon Sep 17 00:00:00 2001
From: Yang Tse <yangsita@gmail.com>
Date: Sat, 29 Oct 2011 14:58:50 +0200
Subject: [PATCH 2/3] gtls.c: gnutls_transport_set_global_errno() deprecated
 in version 2.12.3 (cherry picked from commit
 f5bb37018647f6088398ca127235ce776eec9bbe)

---
 lib/gtls.c |   13 +++++++++++--
 1 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/lib/gtls.c b/lib/gtls.c
index 7ca46c8..c1e9cae 100644
--- a/lib/gtls.c
+++ b/lib/gtls.c
@@ -78,6 +78,17 @@ static void tls_log_func(int level, const char *str)
 #endif
 static bool gtls_inited = FALSE;
 
+#if defined(GNUTLS_VERSION_NUMBER)
+#  if (GNUTLS_VERSION_NUMBER >= 0x020c00)
+#    undef gnutls_transport_set_lowat
+#    define gnutls_transport_set_lowat(A,B) Curl_nop_stmt
+#  endif
+#  if (GNUTLS_VERSION_NUMBER >= 0x020c03)
+#    undef gnutls_transport_set_global_errno
+#    define gnutls_transport_set_global_errno(A) SET_ERRNO((A))
+#  endif
+#endif
+
 /*
  * Custom push and pull callback functions used by GNU TLS to read and write
  * to the socket.  These functions are simple wrappers to send() and recv()
@@ -476,10 +487,8 @@ gtls_connect_step1(struct connectdata *conn,
   gnutls_transport_set_push_function(session, Curl_gtls_push);
   gnutls_transport_set_pull_function(session, Curl_gtls_pull);
 
-#if GNUTLS_VERSION_NUMBER < 0x020c00
   /* lowat must be set to zero when using custom push and pull functions. */
   gnutls_transport_set_lowat(session, 0);
-#endif
 
   /* This might be a reconnect, so we check for a session ID in the cache
      to speed up things */
-- 
1.7.7


From 40ebff7ac932dd2c920d02c8a7aa9da584eafda1 Mon Sep 17 00:00:00 2001
From: Volker Grabsch <vog@notjusthosting.com>
Date: Fri, 28 Oct 2011 13:48:04 +0200
Subject: [PATCH 3/3] static linking for mingw-cross-env

---
 include/curl/curlbuild.h.in |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/include/curl/curlbuild.h.in b/include/curl/curlbuild.h.in
index fe348f4..c428273 100644
--- a/include/curl/curlbuild.h.in
+++ b/include/curl/curlbuild.h.in
@@ -111,6 +111,9 @@
 /*  EXTERNAL INTERFACE SETTINGS FOR CONFIGURE CAPABLE SYSTEMS ONLY  */
 /* ================================================================ */
 
+/* Configure process defines this to 1 when static linking is requested. */
+#undef CURL_STATICLIB
+
 /* Configure process defines this to 1 when it finds out that system  */
 /* header file ws2tcpip.h must be included by the external interface. */
 #undef CURL_PULL_WS2TCPIP_H
-- 
1.7.7

